### PR TITLE
Use a JSON serializer for serializing strings to errorlog file

### LINKED ISSUE #3337: errorlog JSON output should escape backslashes
JSON format requires backslashes to be escaped, I noticed that errorlog JSON output contains unescaped backslashes, i.e. issue location uris containing Windows absolute paths.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #3337 (errorlog JSON output should escape backslashes)

### CODE REVIEW

**Comment by mavasani** [comment]
@tmeschter @srivatsn @JohnHamby @heejaechang can you please review? it should be a quick one.

**Comment by heejaechang** [comment]
:+1:

File: src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs [hunk 1]
```diff
@@ -272,7 +275,8 @@ private void WriteValue(Value value, bool isFirst = true, bool valueInList = fal
 
         private void WriteValue(string value)
         {
-            _writer.Write($"\"{value}\"");
+            _writer.Flush();
+            s_jsonStringSerializer.WriteObject(_writer.BaseStream, value);
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs]: is this exception or thread safe?

Comment (heejaechang) [inline src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs]: using writer with the serializer might be cheaper. WriteObject seems create writer on every call.

Comment (mavasani) [inline src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs]: It is not exception safe, but the values output to the error log file should be generated by the compiler, so shouldn't cause an exception.

Thread safety: You are correct it is not thread safe, and can be a problem in VBCSCompiler mode. I'll switch this to an instance field, and also associate the instance to the writer itself (as per your last review comment).

Comment (mavasani) [inline src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs]: @heejaechang The DataContractJsonSerializer.WriteObject overload that takes a writer is not supported on portable (see [here](https://msdn.microsoft.com/en-us/library/system.runtime.serialization.json.datacontractjsonserializer.writeobject(v=vs.110).aspx)).

For now, I am just going to make this an instance field to address the thread safety.

**Comment by mavasani** [comment]
@heejaechang I have addressed your feedback.

**Comment by heejaechang** [comment]
@mavasani ya, I know. I already gave :+1:

**Pull request merged**